File: Sources/HerdrKit/Transcript/ChatMessage.swift
import Foundation

/// A single WhatsApp-style bubble derived from a Claude Code transcript turn.
/// One transcript entry (a user or assistant turn) maps to one ChatMessage; the
/// turn's parts become ordered `segments` so the UI can show text while
/// collapsing thinking and tool activity.
public struct ChatMessage: Sendable, Identifiable, Equatable {
    public enum Role: String, Sendable {
        case user
        case assistant
        case system
    }

    public let id: String
    public let role: Role
    public let segments: [MessageSegment]
    public let timestamp: Date?
    /// Which agent produced this (e.g. "claude"). Set when a workspace thread
    /// merges more than one agent so bubbles can be labelled.
    public let agentLabel: String?
    /// Sidechain = subagent chatter. Kept but flagged so the UI can hide it.
    public let isSidechain: Bool

    public init(
        id: String,
        role: Role,
        segments: [MessageSegment],
        timestamp: Date?,
        agentLabel: String? = nil,
        isSidechain: Bool = false
    ) {
        self.id = id
        self.role = role
        self.segments = segments
        self.timestamp = timestamp
        self.agentLabel = agentLabel
        self.isSidechain = isSidechain
    }

    /// The plain-text a chat bubble shows (text segments joined). Empty when the
    /// turn was pure thinking/tool activity.
    public var displayText: String {
        segments.compactMap { segment in
            if case let .text(value) = segment { return value }
            return nil
        }
        .joined(separator: "\n")
    }

    /// True when the turn carried no user-visible text (only thinking/tools).
    public var isToolOnly: Bool {
        displayText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

public enum MessageSegment: Sendable, Equatable {
    case text(String)
    case thinking(String)
    case toolUse(name: String, input: String?)
    case toolResult(String)
}

File: android/app/src/main/java/dev/herdr/herdrchat/core/transcript/ChatMessage.kt
package dev.herdr.herdrchat.core.transcript

import kotlinx.serialization.Serializable

/** A single WhatsApp-style bubble derived from a Claude Code transcript turn.
 *  One transcript entry maps to one ChatMessage; the turn's parts become ordered
 *  [segments] so the UI can show text while collapsing thinking and tool activity.
 *  Serializable so the thread cache can persist parsed history to disk. */
@Serializable
data class ChatMessage(
    val id: String,
    val role: Role,
    val segments: List<MessageSegment>,
    val timestamp: Long?,          // epoch millis, or null
    /** Which agent produced this (e.g. "claude"); set when a workspace thread
     *  merges more than one agent so bubbles can be labelled. */
    val agentLabel: String? = null,
    /** Sidechain = subagent chatter. Kept but flagged so the UI can hide it. */
    val isSidechain: Boolean = false,
) {
    enum class Role { USER, ASSISTANT, SYSTEM }

    /** The plain text a bubble shows (text segments joined). Empty when the turn
     *  was pure thinking/tool activity. */
    val displayText: String
        get() = segments.filterIsInstance<MessageSegment.Text>().joinToString("\n") { it.value }

    /** True when the turn carried no user-visible text (only thinking/tools). */
    val isToolOnly: Boolean get() = displayText.isBlank()
}

@Serializable
sealed interface MessageSegment {
    @Serializable
    data class Text(val value: String) : MessageSegment
    @Serializable
    data class Thinking(val value: String) : MessageSegment
    @Serializable
    data class ToolUse(val name: String, val input: String?) : MessageSegment
    @Serializable
    data class ToolResult(val value: String) : MessageSegment
}
